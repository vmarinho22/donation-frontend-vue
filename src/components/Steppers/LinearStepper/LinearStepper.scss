@use '../../../styles/media-queries.scss' as *;
@use '../../../styles/media-theme.scss' as *;
@import '../../../styles/colors.scss';

.linear-stepper {
  .stepper-dots {
    .container {
      display: flex;
      align-items: center;
      justify-items: center;
      flex-direction: column;
      gap: 8px;
      
      .item {
        display: flex;
        justify-content: center;
        align-items: center;
        width: 48px;
        height: 48px;
        border-radius: 50%;
        background-color: $default-gray-200;
        cursor: pointer;
        transition: 0.5s ease-in-out;

        &:hover {
          opacity: 0.7;
        }

        &.selected {
          background-color: $blood-red-500;
          color: #fff;
        }

        @include xs {
          width: 36px;
          height: 36px;
        }

        @include onDark {
          color: #000;
        }
      }
    }
  }
}
