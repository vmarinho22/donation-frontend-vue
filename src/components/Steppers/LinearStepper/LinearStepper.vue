<template>
  <v-container fluid class="linear-stepper">
    <v-row class="stepper-dots" align="center" justify="space-between" style="margin-bottom: 8px">
      <div class="container" v-for="(step, index) in steps" :key="step.title">
        <p class="item" :class="currentStep === index && 'selected'" @click="setStep(index)">
          {{ index + 1 }}
        </p>
        <p>{{ step.title }}</p>
      </div>
    </v-row>
    <v-row align="center" justify="center" no-gutters>
      <component :is="component" v-bind="steps[currentStep].props" />
    </v-row>
    <v-row align="center" justify="end" no-gutters>
      <v-btn variant="text" :disabled="currentStep === 0" @click="back">
        {{ $t('buttons.back') }}
      </v-btn>
      <v-btn :color="Colors.bloodRed[500]" @click="advance">
        {{ currentStep === steps.length - 1 ? $t('buttons.finish') : $t('buttons.advance') }}
      </v-btn>
    </v-row>
  </v-container>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue'
import type { Component } from 'vue'
import { Colors } from '@/constants/colors'

type LinearStepperProps = {
  title: string
  component: Component,
  props?: Record<string, unknown>
}

const { steps, onFinish } = defineProps<{
  steps: LinearStepperProps[];
  onFinish: () => void;
}>()

const currentStep = ref(0)

const component = computed<Component>(() => steps[currentStep.value].component)

function setStep(index: number) {
  currentStep.value = index
}

function advance() {
  if (currentStep.value < steps.length - 1) {
    currentStep.value += 1
  } else if (currentStep.value === steps.length - 1) {
    onFinish();
  }
}

function back() {
  if (currentStep.value > 0) {
    currentStep.value -= 1
  }
}
</script>

<style lang="scss" src="./LinearStepper.scss" scoped />
