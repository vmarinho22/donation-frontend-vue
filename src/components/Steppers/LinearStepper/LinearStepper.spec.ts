import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import LinearStepper from './LinearStepper.vue'
import vuetify from '@/tests/plugins/vuetify'
import MockComponentVue from '@/tests/components/MockComponent.vue'

const steps = [
  {
    title: 'Step 1',
    component: MockComponentVue,
    props: {
      showPasswordInput: true
    }
  },
  {
    title: 'Step 2',
    component: MockComponentVue
  },
  {
    title: 'Step 2',
    component: MockComponentVue
  }
]

describe('LinearStepper', () => {
  it('renders the component', () => {
    const wrapper = mount(LinearStepper, {
      global: {
        plugins: [vuetify]
      },
      props: {
        steps,
        onFinish: async () => {}
      }
    })
    expect(wrapper.html()).toMatchSnapshot();
  })
})