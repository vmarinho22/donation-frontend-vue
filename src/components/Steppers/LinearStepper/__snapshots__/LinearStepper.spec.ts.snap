// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LinearStepper > renders the component 1`] = `
"<div data-v-b7d7fd08=\\"\\" class=\\"v-container v-container--fluid v-locale--is-ltr linear-stepper\\">
  <div data-v-b7d7fd08=\\"\\" class=\\"v-row align-center justify-space-between stepper-dots\\" style=\\"margin-bottom: 8px;\\">
    <div data-v-b7d7fd08=\\"\\" class=\\"container\\">
      <p data-v-b7d7fd08=\\"\\" class=\\"item selected\\">1</p>
      <p data-v-b7d7fd08=\\"\\">Step 1</p>
    </div>
    <div data-v-b7d7fd08=\\"\\" class=\\"container\\">
      <p data-v-b7d7fd08=\\"\\" class=\\"item\\">2</p>
      <p data-v-b7d7fd08=\\"\\">Step 2</p>
    </div>
    <div data-v-b7d7fd08=\\"\\" class=\\"container\\">
      <p data-v-b7d7fd08=\\"\\" class=\\"item\\">3</p>
      <p data-v-b7d7fd08=\\"\\">Step 2</p>
    </div>
  </div>
  <div data-v-b7d7fd08=\\"\\" class=\\"v-row v-row--no-gutters align-center justify-center\\">
    <p data-v-b7d7fd08=\\"\\" showpasswordinput=\\"true\\">Mock component</p>
  </div>
  <div data-v-b7d7fd08=\\"\\" class=\\"v-row v-row--no-gutters align-center justify-end\\"><button data-v-b7d7fd08=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--disabled v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text\\" disabled=\\"\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
      <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\">buttons.back</span>
      <!---->
      <!---->
    </button><button data-v-b7d7fd08=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated\\" style=\\"background-color: rgb(201, 10, 0); color: rgb(255, 255, 255); caret-color: #fff;\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
      <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\">buttons.advance</span>
      <!---->
      <!---->
    </button></div>
</div>"
`;
